subsys/usb/class/audio/audio.c

```diff
@@ -199,10 +199,48 @@ void recv_callback(u8_t ep, enum usb_dc_ep_cb_status_code status)
 
 	dev_data = CONTAINER_OF(common, struct usb_audio_dev_data_t, common);
 
-	ret = usb_read(ep, dev_data->rx_buf, AUDIO_RX_BUFFER_SIZE, &ret_bytes);
+	buffer = net_buf_alloc_fixed(&audio_data_pool, K_NO_WAIT);
+	if (!buffer) {
+		LOG_ERR("Failed to allocate data buffer, %p", buffer);
+		return;
+	}
+
+	ret = usb_read(ep, buffer->data, AUDIO_RX_BUFFER_SIZE, &ret_bytes);
 	if (ret) {
 		LOG_ERR("ret=%d ", ret);
+		net_buf_unref(buffer);
+		return;
 	}
+
+	if ((!ret_bytes) || (!dev_data->ops && !dev_data->ops->data_recv_cb)) {
+		LOG_ERR("No callback/No data, ret_bytes=%d", ret_bytes);
+		net_buf_unref(buffer);
+		return;
+	}
+
+	dev_data->ops->data_recv_cb(ep, buffer, ret_bytes);
+}
+
+void usb_audio_headphones_register(struct device *dev,
```

Comment:
Discussed offline:
There will be a  `_register()` function for each common device (headphones, microphone, headset) and one generic. Those will add a device specific union/pointer inside `usb_audio_dev_data_t`.
It would be nice to have a `usb_audio_device_init()` function that is used by all and could handle both predefined structures and user-defined ones in generic case.